moveit_ros/perception/handeye_calibration_target/src/handeye_aruco_target.cpp

```diff
@@ -0,0 +1,282 @@
+    cv::aruco::detectMarkers(image, dictionary, marker_corners, marker_ids, params_ptr);
+    if (marker_ids.empty())
+    {
+      ROS_DEBUG_STREAM_NAMED(LOGNAME, "No aruco marker detected.");
+      return false;
+    }
+
+    // Refine markers borders
+    std::vector<std::vector<cv::Point2f>> rejected_corners;
+    cv::aruco::refineDetectedMarkers(image, board, marker_corners, marker_ids, rejected_corners, camera_matrix_,
+                                     distortion_coeffs_);
+
+    // Estimate aruco board pose
+    int valid = cv::aruco::estimatePoseBoard(marker_corners, marker_ids, board, camera_matrix_, distortion_coeffs_,
+                                             rotation_vect_, translation_vect_);
+
+    // Draw the markers and frame axis if at least one marker is detected
+    if (valid == 0)
+    {
+      ROS_WARN_STREAM_NAMED(LOGNAME, "Cannot estimate aruco board pose.");
+      return false;
+    }
+
+    if (std::log10(std::fabs(rotation_vect_[0])) > 10 || std::log10(std::fabs(rotation_vect_[1])) > 10 ||
+        std::log10(std::fabs(rotation_vect_[2])) > 10 || std::log10(std::fabs(translation_vect_[0])) > 10 ||
+        std::log10(std::fabs(translation_vect_[1])) > 10 || std::log10(std::fabs(translation_vect_[2])) > 10)
+    {
+      ROS_WARN_STREAM_NAMED(LOGNAME, "Invalid target pose, please check CameraInfo msg.");
+      return false;
+    }
+
+    cv::Mat image_rgb;
+    cv::cvtColor(image, image_rgb, cv::COLOR_GRAY2RGB);
+    cv::aruco::drawDetectedMarkers(image_rgb, marker_corners);
+    drawAxis(image_rgb, camera_matrix_, distortion_coeffs_, rotation_vect_, translation_vect_, 0.1);
+    image = image_rgb;
+  }
+  catch (const cv::Exception& e)
+  {
+    ROS_ERROR_STREAM_NAMED(LOGNAME, "Aruco target detection exception: " << e.what());
+    return false;
+  }
+
+  return true;
+}
+
+geometry_msgs::TransformStamped HandEyeArucoTarget::getTransformStamped(const std::string& frame_id) const
+{
+  geometry_msgs::TransformStamped transform_stamped;
+  transform_stamped.header.stamp = ros::Time::now();
+  transform_stamped.header.frame_id = frame_id;
+  transform_stamped.child_frame_id = "handeye_target";
+
+  tf2::Quaternion quaternion(0, 0, 0, 1);
+  convertToTFQuaternion(rotation_vect_, quaternion);
+  transform_stamped.transform.rotation.x = quaternion.getX();
+  transform_stamped.transform.rotation.y = quaternion.getY();
+  transform_stamped.transform.rotation.z = quaternion.getZ();
+  transform_stamped.transform.rotation.w = quaternion.getW();
+
+  std::vector<double> translation(3, 0);
+  convertToStdVector(translation_vect_, translation);
+  transform_stamped.transform.translation.x = translation[0];
+  transform_stamped.transform.translation.y = translation[1];
+  transform_stamped.transform.translation.z = translation[2];
+
+  return transform_stamped;
+}
+
+bool HandEyeArucoTarget::convertToTFQuaternion(const cv::Vec3d& input_rvect, tf2::Quaternion& quaternion) const
```

Comment:
What you could do here, is to modify the function to log/print out something useful and return false if something went wrong. Asserts help for debug builds but for release builds they are just removed altogether. The point I'm trying to make here is that returning boolean but then setting the return path to always return true is not ideal. I have nothing against asserts though.

Reply:
Actually I struggled here to choose an assert or an error report. Previously I used the error report and return false. [Here](https://docs.opencv.org/3.4/db/de0/group__core__utils.html#gaf62bcd90f70e275191ab95136d85906b) claims that `CV_assert` also checks for release configurations. Seems ` CV_DbgAssert` only checks in debug build. But I think it's better to use an error report here that can provide more information.

Reply:
Ah, sorry about the misleading comment. Should have checked the cv_assert documentation but relied on memory instead. As for the changes, I think it's really good now.